### PR TITLE
[RLlib] Fix performance and functionality flaws in attention nets (via Trajectory view API).

### LINKED ISSUE #10883: [RLlib Trajectory View API] Trajectory View API works with our Attention Nets.
The Trajectory View API works with our Attention Nets.
- This may involve enhancements on the attention networks itself (try to cache previous inputs (older observations) and successive computations on these for better performance).

### PR DESCRIPTION
**Pull request by sven1977**
- RLlib's attention nets (GTrXL) have been forced so far to run "inside" RLlib's RNN API (previous internal states are being passed as new state-ins in subsequent timesteps). This is not favorable for attention nets, which need a different handling and time-slicing of past states (attention net's memory). The trajectory view API allows for specifying the needed time-step ranges for forward passes and batched train passes through attention nets.
- Besides the above, the handling of the tau-memory of attention nets was also not correct. This PR fixes existing bugs.
- In a follow up PR, the torch version of GTrXL will be fully included in the testing as well (to make sure it's 100% en-par with the tf version).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 1]
```diff
@@ -47,7 +48,8 @@ def ppo_surrogate_loss(
 
     # RNN case: Mask away 0-padded chunks at end of time axis.
     if state:
-        max_seq_len = tf.reduce_max(train_batch["seq_lens"])
+        B = tf.shape(train_batch["seq_lens"])[0]
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:51]: This was not safe for attention nets, where the max seq len must be kept (attention nets are not flexible wrt the seq-len as RNNs are).

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 2]
```diff
@@ -188,13 +190,16 @@ def postprocess_ppo_gae(
...
     # Trajectory has been truncated -> last r=VF estimate of last obs.
     else:
-        next_state = []
-        for i in range(policy.num_state_tensors()):
-            next_state.append(sample_batch["state_out_{}".format(i)][-1])
-        last_r = policy._value(sample_batch[SampleBatch.NEXT_OBS][-1],
-                               sample_batch[SampleBatch.ACTIONS][-1],
-                               sample_batch[SampleBatch.REWARDS][-1],
-                               *next_state)
+        if policy.config["_use_trajectory_view_api"]:
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:193]: Eventually remove boilerplate input-dict generation for non-traj. view case (in else block).

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 3]
```diff
@@ -286,20 +291,32 @@ def __init__(self, obs_space, action_space, config):
...
-                    SampleBatch.PREV_ACTIONS: tf.convert_to_tensor(
-                        [prev_action]),
-                    SampleBatch.PREV_REWARDS: tf.convert_to_tensor(
-                        [prev_reward]),
-                    "is_training": tf.convert_to_tensor([False]),
-                }, [tf.convert_to_tensor([s]) for s in state],
-                                          tf.convert_to_tensor([1]))
-                # [0] = remove the batch dim.
-                return self.model.value_function()[0]
+            # Traj. view API.
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:294]: Same as above. Use input_dict instead, generated fully automatically by traj. view API according to view requirements for this policy/model.

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 4]
```diff
@@ -344,6 +361,15 @@ def setup_mixins(policy: Policy, obs_space: gym.spaces.Space,
     LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
 
 
+def view_requirements_fn(policy):
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_tf_policy.py:364]: PPO and others need a complete input dict "view" for calculating GAE at the end of the trajectory. This view requirement sets this up.

Comment (sven1977) [reply rllib/agents/ppo/ppo_tf_policy.py:369]: (see `postprocess_trajectory` above)

File: rllib/agents/ppo/tests/test_appo.py [hunk 5]
```diff
@@ -24,6 +24,7 @@ def test_appo_compilation(self):
         for _ in framework_iterator(config):
             print("w/o v-trace")
             _config = config.copy()
+            _config["vtrace"] = False
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_appo.py:27]: this was a bug

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 6]
```diff
@@ -45,24 +49,30 @@ class _AgentCollector:
...
         # The simple timestep count for this agent. Gets increased by one
         # each time a (non-initial!) observation is added.
         self.count = 0
 
-    def add_init_obs(self, episode_id: EpisodeID, agent_id: AgentID,
-                     env_id: EnvID, init_obs: TensorType,
+    def add_init_obs(self, episode_id: EpisodeID, agent_index: int,
+                     env_id: EnvID, t: int, init_obs: TensorType,
                      view_requirements: Dict[str, ViewRequirement]) -> None:
         """Adds an initial observation (after reset) to the Agent's trajectory.
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:66]: Change the description. It adds more than a single observation.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 7]
```diff
@@ -112,49 +128,91 @@ def build(self, view_requirements: Dict[str, ViewRequirement]) -> \
...
+
+            # Range of indices on time-axis, make sure to create
+            if view_req.data_rel_pos_from is not None:
+                if view_req.batch_repeat_value > 1:
+                    count = int(
+                        math.ceil((len(np_data[data_col]) - self.shift_before)
+                                  / view_req.batch_repeat_value))
+                    repeat_count = (view_req.data_rel_pos_to -
+                                    view_req.data_rel_pos_from + 1)
+                    data = np.asarray([
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:176]: Same as above, big confused. Add comments on what these lines of code do

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 8]
```diff
@@ -112,49 +128,91 @@ def build(self, view_requirements: Dict[str, ViewRequirement]) -> \
...
+            elif isinstance(view_req.data_rel_pos, np.ndarray):
+                data = np_data[data_col][self.shift_before + obs_shift +
+                                         view_req.data_rel_pos]
+            # Single index.
             else:
-                data = np_data[data_col][self.shift_before + shift:shift]
+                shift = view_req.data_rel_pos + obs_shift
+                # Shift is exactly 0: Use trajectory as is.
+                if shift == 0:
+                    data = np_data[data_col][self.shift_before:]
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:198]: Same as before

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 9]
```diff
@@ -208,6 +269,43 @@ def _build_buffers(self, single_row: Dict[str, TensorType]) -> None:
                         [np.zeros(shape=shape, dtype=dtype)
                          for _ in range(shift)]
 
+    def _get_input_dict(self, view_reqs, abs_pos: int = -1) -> \
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:272]: Add description what this method does

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 10]
```diff
@@ -245,9 +350,17 @@ def add_postprocessed_batch_for_training(
...
             self.buffers[view_col].extend(data)
         # Add the agent's trajectory length to our count.
         self.count += batch.count
+        # Adjust the seq-lens array depending on the incoming agent sequences.
+        if self.seq_lens is not None:
+            max_seq_len = self.policy.config["model"]["max_seq_len"]
+            count = batch.count
+            while count > 0:
+                self.seq_lens.append(min(count, max_seq_len))
+                count -= max_seq_len
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:363]: Where is this count used?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 11]
```diff
@@ -259,20 +372,22 @@ def build(self):
                 this policy.
         """
         # Create batch from our buffers.
-        batch = SampleBatch(self.buffers)
-        assert SampleBatch.UNROLL_ID in batch.data
+        batch = SampleBatch(
+            self.buffers, _seq_lens=self.seq_lens, _dont_check_lens=True)
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:376]: What is _dont_check_lens?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 12]
```diff
@@ -259,20 +372,22 @@ def build(self):
...
         self.buffers.clear()
-        # Reset count to 0.
+        # Reset count to 0 and seq-lens to empty list.
         self.count = 0
+        if self.seq_lens is not None:
+            self.seq_lens = []
         return batch
 
 
 class _PolicyCollectorGroup:
```
Comment (michaelzhiluo) [inline rllib/evaluation/collectors/simple_list_collector.py:386]: Probably add comments on what this is

File: rllib/policy/sample_batch.py [hunk 13]
```diff
@@ -245,12 +252,30 @@ def slice(self, start: int, end: int) -> "SampleBatch":
...
+        if self.seq_lens is not None and len(self.seq_lens) > 0:
+            data = {k: v[start:end] for k, v in self.data.items()}
+            # Fix state_in_x data.
+            count = 0
+            state_start = None
+            seq_lens = None
+            for i, seq_len in enumerate(self.seq_lens):
+                count += seq_len
+                if count >= end:
+                    data["state_in_0"] = self.data["state_in_0"][state_start:
```
Comment (michaelzhiluo) [inline rllib/policy/sample_batch.py:264]: Add comment on what this does

File: rllib/policy/policy.py [hunk 14]
```diff
@@ -705,14 +695,41 @@ def _get_dummy_batch_from_view_requirements(
...
                 ret[view_col] = \
                     np.zeros((batch_size, ) + shape[1:], np.float32)
             else:
-                ret[view_col] = np.zeros_like(
-                    [view_req.space.sample() for _ in range(batch_size)])
-        return SampleBatch(ret)
+                # Range of indices on time-axis, make sure to create
+                if view_req.data_rel_pos_from is not None:
+                    ret[view_col] = np.zeros_like([[
+                        view_req.space.sample()
```
Comment (michaelzhiluo) [inline rllib/policy/policy.py:709]: Same add comment here

File: rllib/models/tf/layers/relative_multi_head_attention.py [hunk 15]
```diff
@@ -121,3 +125,14 @@ def rel_shift(x: TensorType) -> TensorType:
         x = tf.reshape(x, x_size)
 
         return x
+
+
+class PositionalEmbedding(tf.keras.layers.Layer):
```
Comment (michaelzhiluo) [inline rllib/models/tf/layers/relative_multi_head_attention.py:130]: Add comments on what this does (how it initializes embedding per position based on cos/sin something)

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 6 above)
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py]: No, it doesn't it's really just adds a single one. Same as it used to work w/ SampleBatchBuilder.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 16]
```diff
@@ -45,24 +49,30 @@ class _AgentCollector:
 
     _next_unroll_id = 0  # disambiguates unrolls within a single episode
 
-    def __init__(self, shift_before: int = 0):
-        self.shift_before = max(shift_before, 1)
+    def __init__(self, view_reqs):
+        self.shift_before = -min(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:53]: done

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 7 above)
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:164]: done

Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:164]: Provided an example.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 8 above)
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:187]: Provided an example.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 11 above)
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:331]: added explanation.

**Comment by sven1977** [comment]
Moved here:
https://github.com/ray-project/ray/pull/12753

**Pull request closed without merging**